### PR TITLE
swift-lang: new recipe

### PR DESCRIPTION
**Pull request by return**
This port of Swift 3.1 will only work on x86_64 and requires installing llvm-4.0.0-2 because of this [commit](https://github.com/return/haikuports/commit/b6316842ed1d62ccb0a12d8f7353edea96e536bf). 

Further progress and documentation of the swift port can be found on the [swift-haiku-build](https://github.com/return/swift-haiku-build) repository.

What works so far:
* Running Swift programs through the interpreter. 
* Importing libraries. (Passing -sdk '/' to compiler or export SDKROOT="/")
* SIL/IR Code Generation.
* Emitting object code.
* Developer tools.
* Generating executables with swiftc
 
(UPDATE 18th June 2017):
Partial support for running swift programs via interpreter.
_(Removing some concurrency related code in the MetadataCache class, allows swift programs to run in the interpreter, but this hack only serves as a proof of concept showing that swift programs can run in Haiku.)_

(UPDATE 6th July 2017):
Compiling swift source files into executables now seem to work.

### CODE REVIEW

File: dev-lang/swift/swift-3.1.recipe [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	llvm${secondaryArchSuffix}_clang
+	devel:libuuid$secondaryArchSuffix
+	devel:libicuuc$secondaryArchSuffix
+	devel:libicui18n$secondaryArchSuffix
+	devel:libicudata$secondaryArchSuffix
+	devel:libexecinfo$secondaryArchSuffix
+	devel:libiconv$secondaryArchSuffix
+	devel:libcharset$secondaryArchSuffix
+	devel:libedit$secondaryArchSuffix
+	sqlite$secondaryArchSuffix_devel
```
Comment (diversys) [inline dev-lang/swift/swift-3.1.recipe]: Should be devel:libsqlite3$secondaryArchSuffix

File: dev-lang/swift/swift-3.1.recipe [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	devel:libncurses$secondaryArchSuffix
+	devel:libxml2$secondaryArchSuffix
+	gcc_syslibs$secondaryArchSuffix_devel
+	devel:libgcc$secondaryArchSuffix
+	cmd:clang
```
Comment (diversys) [inline dev-lang/swift/swift-3.1.recipe]: All cmd: should go to BUILD_PREREQUIRES

File: dev-lang/swift/swift-3.1.recipe [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+	# Build faster by building only for X86
+	python2 ./swift/utils/build-script -R -m '--extra-cmake-options= -DLLVM_ENABLE_ASSERTIONS=TRUE -DCMAKE_SKIP_RPATH=FALSE 
+	 -DLLVM_ENABLE_RTTI=ON -DLLVM_TARGETS_TO_BUILD=X86 -DLLVM_ENABLE_THREADS=NO -DCMAKE_C_FLAGS=-fPIC -DCMAKE_CXX_FLAGS=-fPIC' --stdlib-deployment-targets=haiku-x86_64 -j4 $jobArgs --reconfigure
+}
+
+INSTALL()
+{
+	mkdir -p $binDir
+	mkdir -p $libDir
```
Comment (diversys) [inline dev-lang/swift/swift-3.1.recipe]: You can combine these mkdir -p $binDir $libDir

File: dev-lang/swift/swift-3.1.recipe [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+	lib:libxml2$secondaryArchSuffix
+	lib:libexecinfo$secondaryArchSuffix
+	lib:libiconv$secondaryArchSuffix
+	lib:libcharset$secondaryArchSuffix
+	lib:libz$secondaryArchSuffix
+	sqlite$secondaryArchSuffix
+	"
+BUILD_REQUIRES="
+	haiku${secondaryArchSuffix}_devel
+	llvm${secondaryArchSuffix}
```
Comment (diversys) [inline dev-lang/swift/swift-3.1.recipe]: Already included by cmd:clang$secondaryArchSuffix

File: dev-lang/swift/swift-3.1.recipe [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+	cmd:clang++
+	cmd:python
+	"
+BUILD_PREREQUIRES="
+	llvm${secondaryArchSuffix}
+	llvm${secondaryArchSuffix}_clang
+	gcc_syslibs$secondaryArchSuffix
+	gcc_syslibs_devel$secondaryArchSuffix
+	devel:libgcc$secondaryArchSuffix
+	libtool$secondaryArchSuffix
```
Comment (diversys) [inline dev-lang/swift/swift-3.1.recipe]: Should be cmd:libtoolize$secondaryArchSuffix

File: dev-lang/swift/swift-3.1.recipe [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+	devel:libncurses$secondaryArchSuffix
+	devel:libxml2$secondaryArchSuffix
+	gcc_syslibs$secondaryArchSuffix_devel
+	devel:libgcc$secondaryArchSuffix
+	cmd:clang
+	cmd:clang++
+	cmd:python
+	"
+BUILD_PREREQUIRES="
+	llvm${secondaryArchSuffix}
```
Comment (diversys) [inline dev-lang/swift/swift-3.1.recipe]: You only need braces for haiku${secondaryArchSuffix}_devel I think

File: dev-lang/swift/swift-3.1.recipe [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+	gcc_syslibs$secondaryArchSuffix
+	gcc_syslibs_devel$secondaryArchSuffix
+	devel:libgcc$secondaryArchSuffix
+	libtool$secondaryArchSuffix
+	cmd:gcc$secondaryArchSuffix
+	cmd:clang$secondaryArchSuffix
+	cmd:clang++$secondaryArchSuffix
+	cmd:ld$secondaryArchSuffix
+	cmd:swig
+	cmd:pkg_config
```
Comment (diversys) [inline dev-lang/swift/swift-3.1.recipe]: cmd:pkg_config$secondaryArchSuffix

File: dev-lang/swift/swift-3.1.recipe [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+	cmd:make
+	cmd:python2
+	cmd:python
+	cmd:autoconf
+	cmd:cmake
+	cmd:ninja
+	cmd:sh
```
Comment (diversys) [inline dev-lang/swift/swift-3.1.recipe]: Shouldn't be needed

File: dev-lang/swift_lang/swift_lang-3.1.recipe [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+	# Fix permisssions for build-script.
+	chmod +x ./swift/utils/build-script
+	chmod +x ./swift/utils/build-script-impl
+	chmod -R 755 ./swift/utils
+
+	export HOST_CC="/bin/clang"
+	export HOST_CXX="/bin/clang++"
+	export CXXFLAGS="-fPIC"
+	export CFLAGS="-fPIC"
```
Comment (korli) [inline dev-lang/swift_lang/swift_lang-3.1.recipe]: Is fPIC really needed with the updated clang?

File: dev-lang/swift_lang/swift_lang-3.1.recipe [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+
+	# Build faster by building only for X86
+	python ./swift/utils/build-script -R -m \
+	--extra-cmake-options='-DLLVM_ENABLE_ASSERTIONS=TRUE
```
Comment (korli) [inline dev-lang/swift_lang/swift_lang-3.1.recipe]: Miss a backslash at the end of the line, next four lines too.

File: dev-lang/swift_lang/swift_lang-3.1.recipe [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+	 -DCMAKE_SKIP_RPATH=FALSE -DLLVM_ENABLE_RTTI=ON
+	 -DLLVM_TARGETS_TO_BUILD=X86 -DLLVM_ENABLE_THREADS=NO
+	 -DCMAKE_C_FLAGS=-fPIC -DCMAKE_CXX_FLAGS=-fPIC' \
+	 --stdlib-deployment-targets=haiku-x86_64 -j4 $jobArgs \
```
Comment (korli) [inline dev-lang/swift_lang/swift_lang-3.1.recipe]: Please drop -j4

File: dev-lang/swift_lang/swift_lang-3.1.recipe [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+	devel:libgcc$secondaryArchSuffix
+	cmd:libtool$secondaryArchSuffix
+	cmd:gcc$secondaryArchSuffix
+	cmd:clang$secondaryArchSuffix
+	cmd:clang++$secondaryArchSuffix
+	cmd:ld$secondaryArchSuffix
+	cmd:pkg_config$secondaryArchSuffix
+	cmd:swig
+	cmd:make
+	cmd:python
```
Comment (korli) [inline dev-lang/swift_lang/swift_lang-3.1.recipe]: Should be cmd python2

File: dev-lang/swift_lang/swift_lang-3.1.recipe [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+	cmd:swift_remoteast_test$secondaryArchSuffix = $portVersion
+	cmd:sil_func_extractor$secondaryArchSuffix = $portVersion
+	cmd:sil_opt$secondaryArchSuffix = $portVersion
+	cmd:sil_nm$secondaryArchSuffix = $portVersion
+	cmd:sil_passpipeline_dumper$secondaryArchSuffix = $portVersion
+	cmd:lldb_moduleimport_test$secondaryArchSuffix = $portVersion
+	"
+REQUIRES="
+	haiku$secondaryArchSuffix
+	llvm${secondaryArchSuffix} >= 3.7.0 base
```
Comment (diversys) [inline dev-lang/swift_lang/swift_lang-3.1.recipe]: You can drop brackets here

File: dev-lang/swift_lang/swift_lang-3.1.recipe [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+	"
+
+BUILD()
+{
+	cd ..
+
+	mkdir -p swift
+	mkdir -p clang
+	mkdir -p llvm
+	mkdir -p cmark
```
Comment (diversys) [inline dev-lang/swift_lang/swift_lang-3.1.recipe]: mkdir -p swift clang llvm cmark

File: dev-lang/swift_lang/swift_lang-3.1.recipe [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+
+	# Copy all the source files
+
```
Comment (diversys) [inline dev-lang/swift_lang/swift_lang-3.1.recipe]: drop the new line

File: dev-lang/swift_lang/swift_lang-3.1.recipe (hunk 9 above)
Comment (return) [inline dev-lang/swift_lang/swift_lang-3.1.recipe]: Building the toolchain without the PIE flag isn't possible. It throws symbol relocation errors whilst linking some object files. This [gist](https://gist.github.com/return/d04471f208873d8a07b005ebb6290e4b) outline what happens when building without this flag.

File: dev-lang/swift_lang/swift_lang-3.1.recipe [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+	devel:libcharset$secondaryArchSuffix
+	devel:libedit$secondaryArchSuffix
+	devel:libsqlite3$secondaryArchSuffix
+	devel:libncurses$secondaryArchSuffix
+	devel:libxml2$secondaryArchSuffix
+	gcc_syslibs$secondaryArchSuffix_devel
+	devel:libgcc$secondaryArchSuffix
+	"
+BUILD_PREREQUIRES="
+	cmd:libtool$secondaryArchSuffix
```
Comment (diversys) [inline dev-lang/swift_lang/swift_lang-3.1.recipe]: Please sort all requires alphabetically

File: dev-lang/swift_lang/swift_lang-3.1.recipe [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+	chmod -R 755 ./swift/utils
+
+	export HOST_CC="/bin/clang"
+	export HOST_CXX="/bin/clang++"
+
+	# Build faster by building only for X86
+	python2 ./swift/utils/build-script -R -m --extra-cmake-options='-DLLVM_ENABLE_ASSERTIONS=TRUE \
+	 -DCMAKE_SKIP_RPATH=FALSE -DLLVM_ENABLE_RTTI=ON -DLLVM_TARGETS_TO_BUILD=X86 \
+	 -DLLVM_ENABLE_THREADS=NO -DCMAKE_C_FLAGS=-fPIC -DCMAKE_CXX_FLAGS=-fPIC' \
+	 --stdlib-deployment-targets=haiku-x86_64 -j4 $jobArgs --reconfigure
```
Comment (diversys) [inline dev-lang/swift_lang/swift_lang-3.1.recipe]: drop -j4

File: dev-lang/swift_lang/swift_lang-3.1.recipe [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+	# Fix permisssions for build-script.
+	chmod +x ./swift/utils/build-script
+	chmod +x ./swift/utils/build-script-impl
+	chmod -R 755 ./swift/utils
+
+	export HOST_CC="/bin/clang"
+	export HOST_CXX="/bin/clang++"
+
+	# Build faster by building only for X86
+	python2 ./swift/utils/build-script -R -m --extra-cmake-options='-DLLVM_ENABLE_ASSERTIONS=TRUE \
```
Comment (diversys) [inline dev-lang/swift_lang/swift_lang-3.1.recipe]: Wrap at 80th character

File: dev-lang/swift_lang/swift_lang-3.1.recipe [hunk 19]
```diff
@@ -0,0 +1,145 @@
+SUMMARY="A modern high performance safe general-purpose programming language"
+DESCRIPTION="Swift is a general-purpose programming language built using a \
+modern approach to safety, performance, and software design patterns. \
+The goal of the Swift project is to create the best available \
+language for uses ranging from systems programming to mobile \
+and desktop apps, scaling up to cloud services. Most importantly, \
+Swift  is designed to make writing and maintaining correct programs \
+easier for the developer.
+"
```
Comment (diversys) [inline dev-lang/swift_lang/swift_lang-3.1.recipe]: Move quote the the previous line

File: dev-lang/swift_lang/swift_lang-3.1.recipe [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+HOMEPAGE="https://swift.org"
+LICENSE="Apache v2"
+COPYRIGHT="2015 Apple Inc"
```
Comment (diversys) [inline dev-lang/swift_lang/swift_lang-3.1.recipe]: Website says it's 2017

File: dev-lang/swift_lang/swift_lang-3.1.recipe (hunk 9 above)
Comment (korli) [inline dev-lang/swift_lang/swift_lang-3.1.recipe]: "without the PIE flag", you meant PIC, right?

Comment (return) [inline dev-lang/swift_lang/swift_lang-3.1.recipe]: Oops! yes I meant PIC, sorry for the confusion.

**Comment by return** [comment]
Here is a proof of concept of swift running on Haiku x86_64.
screenshot4

**Comment by scottmc** [comment]
What's the current status of this one?

**Comment by korli** [comment]
@scottmc the recipe still needs to be fixed (some requires look wrong), and probably tested against a recent hrev of Haiku.

**Pull request merged**